reactive/kubernetes_master.py

```diff
@@ -1944,11 +1965,11 @@ def configure_apiserver():
 
     admission_plugins = [
         'PersistentVolumeLabel',
+        'PodSecurityPolicy',
```

Comment:
Yikes, PodSecurityPolicy is a can of worms. Please check out the [upstream docs on Pod Security Policies](https://kubernetes.io/docs/concepts/policy/pod-security-policy/) and in particular, note this warning:

> PodSecurityPolicies are enforced by enabling the admission controller, but doing so without authorizing any policies will prevent any pods from being created in the cluster.

I believe if we are enabling PodSecurityPolicy, then the charm will also need to create one or more PodSecurityPolicies, along with Roles/ClusterRoles and RoleBindings/ClusterRoleBindings to authorize users to use those policies. It's gonna get messy.

Reply:
Yeah, seems to be.  When I set `RBAC,Node`, everything stops working because (I'm guessing) these aren't set.  It's a requirement of the CIS benchmark though.

Reply:
Ah, no, I misspoke, it just took a while to settle after changing to `RBAC,Node`.  It seems to work without these.

Reply:
Are you sure? If I enable PodSecurityPolicy on a 1.18 cluster with:

```
juju config kubernetes-master api-extra-args=enable-admission-plugins=PersistentVolumeLabel,PodSecurityPolicy,NodeRestriction
```

At first, everything looks fine:

```
$ kubectl get po --all-namespaces
NAMESPACE                         NAME                                                      READY   STATUS    RESTARTS   AGE
ingress-nginx-kubernetes-worker   default-http-backend-kubernetes-worker-7f969db96f-jpxjh   1/1     Running   0          9m38s
ingress-nginx-kubernetes-worker   nginx-ingress-controller-kubernetes-worker-j9blx          1/1     Running   0          9m30s
kube-system                       coredns-6b59b8bd9f-2gnh5                                  1/1     Running   0          9m54s
kube-system                       kube-state-metrics-69f474f8cb-2dbck                       1/1     Running   0          9m54s
kube-system                       metrics-server-v0.3.6-74c87686d-zwm89                     2/2     Running   0          9m54s
kubernetes-dashboard              dashboard-metrics-scraper-78d587b6b-77g4p                 1/1     Running   0          9m54s
kubernetes-dashboard              kubernetes-dashboard-7b76878768-zjnj4                     1/1     Running   0          9m54s
```

But if I delete one of the pods, it never comes back:

```
$ kubectl delete po -n kubernetes-dashboard kubernetes-dashboard-7b76878768-zjnj4
pod "kubernetes-dashboard-7b76878768-zjnj4" deleted

$ kubectl get po --all-namespaces
NAMESPACE                         NAME                                                      READY   STATUS    RESTARTS   AGE
ingress-nginx-kubernetes-worker   default-http-backend-kubernetes-worker-7f969db96f-jpxjh   1/1     Running   0          10m
ingress-nginx-kubernetes-worker   nginx-ingress-controller-kubernetes-worker-j9blx          1/1     Running   0          10m
kube-system                       coredns-6b59b8bd9f-2gnh5                                  1/1     Running   0          11m
kube-system                       kube-state-metrics-69f474f8cb-2dbck                       1/1     Running   0          11m
kube-system                       metrics-server-v0.3.6-74c87686d-zwm89                     2/2     Running   0          11m
kubernetes-dashboard              dashboard-metrics-scraper-78d587b6b-77g4p                 1/1     Running   0          11m
```

and when I check the kube-controller-manager logs, I see it trying and failing to create the pod:

```
$ journalctl -o cat -u snap.kube-controller-manager.daemon | tail -n 8
I0715 16:09:33.539660    2619 replica_set.go:562] Too few replicas for ReplicaSet kubernetes-dashboard/kubernetes-dashboard-7b76878768, need 1, creating 1
I0715 16:09:33.544646    2619 replica_set.go:587] Slow-start failure. Skipping creation of 1 pods, decrementing expectations for ReplicaSet kubernetes-dashboard/kubernetes-dashboard-7b76878768
E0715 16:09:33.544731    2619 replica_set.go:535] sync "kubernetes-dashboard/kubernetes-dashboard-7b76878768" failed with pods "kubernetes-dashboard-7b76878768-" is forbidden: no providers available to validate pod request
I0715 16:09:33.544773    2619 event.go:278] Event(v1.ObjectReference{Kind:"ReplicaSet", Namespace:"kubernetes-dashboard", Name:"kubernetes-dashboard-7b76878768", UID:"663e530b-8726-4619-88c8-d8374a6226f2", APIVersion:"apps/v1", ResourceVersion:"1878", FieldPath:""}): type: 'Warning' reason: 'FailedCreate' Error creating: pods "kubernetes-dashboard-7b76878768-" is forbidden: no providers available to validate pod request
I0715 16:11:46.857978    2619 replica_set.go:562] Too few replicas for ReplicaSet kubernetes-dashboard/kubernetes-dashboard-7b76878768, need 1, creating 1
I0715 16:11:46.862448    2619 replica_set.go:587] Slow-start failure. Skipping creation of 1 pods, decrementing expectations for ReplicaSet kubernetes-dashboard/kubernetes-dashboard-7b76878768
E0715 16:11:46.862517    2619 replica_set.go:535] sync "kubernetes-dashboard/kubernetes-dashboard-7b76878768" failed with pods "kubernetes-dashboard-7b76878768-" is forbidden: no providers available to validate pod request
I0715 16:11:46.862552    2619 event.go:278] Event(v1.ObjectReference{Kind:"ReplicaSet", Namespace:"kubernetes-dashboard", Name:"kubernetes-dashboard-7b76878768", UID:"663e530b-8726-4619-88c8-d8374a6226f2", APIVersion:"apps/v1", ResourceVersion:"1878", FieldPath:""}): type: 'Warning' reason: 'FailedCreate' Error creating: pods "kubernetes-dashboard-7b76878768-" is forbidden: no providers available to validate pod request
```

Additionally, I am unable to create pods myself:
```
$ cat ubuntu.yaml
apiVersion: v1
kind: Pod
metadata:
  name: ubuntu
  namespace: default
spec:
  containers:
  - image: ubuntu
    command:
      - sleep
      - "3600"
    imagePullPolicy: IfNotPresent
    name: ubuntu
  restartPolicy: Always

$ kubectl apply -f ubuntu.yaml
Error from server (Forbidden): error when creating "ubuntu.yaml": pods "ubuntu" is forbidden: no providers available to validate pod request
```

Reply:
It is working when I set it post default 1.18, but it doesn't if I deploy from scratch, due to the scheduler not being able to talk to the API:
`2020-07-15T16:20:07Z kube-scheduler.daemon[20242]: E0715 16:20:07.401045   20242 reflector.go:178] k8s.io/client-go/informers/factory.go:135: Failed to list *v1.Service: Get http://127.0.0.1:8080/api/v1/services?limit=500&resourceVersion=0: dial tcp 127.0.0.1:8080: connect: connection refused`